dbms/src/Interpreters/Join.cpp

```diff
@@ -1050,23 +1050,23 @@ void Join::insertFromBlock(const Block & block, size_t stream_index)
         for (size_t j = stream_index; j < build_concurrency + stream_index; ++j)
         {
             size_t i = j % build_concurrency;
+            if (!dispatch_blocks[i].rows())
+            {
+                continue;
+            }
+            Blocks blocks_to_spill;
             {
                 std::unique_lock partition_lock(partitions_locks[i]);
-                Block & dispatch_block = dispatch_blocks[i];
-                if (!dispatch_block.rows())
-                {
-                    continue;
-                }
-                insertBlockToBuildPartition(dispatch_block, i);
-                trySpillBuildPartition(i, false);
+                insertBlockToBuildPartition(std::move(dispatch_blocks[i]), i);
                 if (partitions[i].spill)
-                {
-                    continue;
-                }
-
-                stored_block = &(partitions[i].build_partition.blocks.back());
+                    blocks_to_spill = trySpillBuildPartition(i, false);
```

Comment:
It is better to change  trySpillBuildPartition to tryGetBuildPartitionSpillBlocks